pkg/driver/controller.go

```diff
@@ -25,28 +25,279 @@ import (
 	// controllerCaps represents the capability of controller service
 	controllerCaps = []csi.ControllerServiceCapability_RPC_Type{
+		csi.ControllerServiceCapability_RPC_CREATE_DELETE_VOLUME,
 		csi.ControllerServiceCapability_RPC_CREATE_DELETE_SNAPSHOT,
 	}
 )
 
 const (
-	volumeParamsDnsName    = "dnsname"
-	volumeParamsMountName  = "mountname"
-	volumeParamsVolumeType = "volumeType"
+	volumeContextVolumeType                   = "volumeType"
+	volumeContextDnsName                      = "dnsName"
+	volumeContextVolumePath                   = "volumePath"
+	volumeParamsVolumeType                    = "volumeType"
+	volumeParamsKmsKeyId                      = "kmsKeyId"
+	volumeParamsAutomaticBackupRetentionDays  = "automaticBackupRetentionDays"
+	volumeParamsCopyTagsToBackups             = "copyTagsToBackups"
+	volumeParamsCopyTagsToVolumes             = "copyTagsToVolumes"
+	volumeParamsDailyAutomaticBackupStartTime = "dailyAutomaticBackupStartTime"
+	volumeParamsDeploymentType                = "deploymentType"
+	volumeParamsDiskIopsConfiguration         = "diskIopsConfiguration"
+	volumeParamsRootVolumeConfiguration       = "rootVolumeConfiguration"
+	volumeParamsThroughputCapacity            = "throughputCapacity"
+	volumeParamsWeeklyMaintenanceStartTime    = "weeklyMaintenanceStartTime"
+	volumeParamsSecurityGroupIds              = "securityGroupIds"
+	volumeParamsSubnetIds                     = "subnetIds"
+	volumeParamsTags                          = "tags"
+	volumeParamsName                          = "name"
+	volumeParamsCopyTagsToSnapshots           = "copyTagsToSnapshots"
+	volumeParamsDataCompressionType           = "dataCompressionType"
+	volumeParamsNfsExports                    = "NfsExports"
+	volumeParamsParentVolumeId                = "parentVolumeId"
+	volumeParamsReadOnly                      = "readOnly"
+	volumeParamsRecordSizeKiB                 = "recordSizeKiB"
+	volumeParamsUserAndGroupQuotas            = "userAndGroupQuotas"
+)
+
+const (
+	fsVolumeType     = "filesystem"
+	volVolumeType    = "volume"
+	rootVolumePath   = "fsx"
+	filesystemPrefix = "fs"
+	volumePrefix     = "fsvol"
 )
 
 func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	klog.V(4).Infof("CreateVolume: called with args %#v", req)
+
+	volName := req.GetName()
+	if len(volName) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume name not provided")
+	}
+
+	volCaps := req.GetVolumeCapabilities()
+	if len(volCaps) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities not provided")
+	}
+
+	if !d.isValidVolumeCapabilities(volCaps) {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities not supported")
+	}
+
+	volumeParams := req.GetParameters()
+	volumeType := volumeParams[volumeParamsVolumeType]
+
+	if volumeType == "filesystem" {
+		storageCapacity := cloud.DefaultFileSystemSize
```

Comment:
If the capacity range is nil, shouldn't the RPC throw an error? Why do we set a default here?